onnx/backend/test/case/node/matmulnbits.py

```diff
@@ -0,0 +1,513 @@
+# Copyright (c) ONNX Project Contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import numpy as np
+import math
+
+import onnx
+from onnx.backend.test.case.base import Base
+from onnx.backend.test.case.node import expect
+from typing import Tuple
+
+def matmulnbits_unpack_zero_points(
+    zero_points: np.ndarray,
+    N: int, n_blocks_per_col: int,
+    bits: int
+) -> np.ndarray:
+    unpacked_zp = []
+    zp_bits_per_n = math.ceil(n_blocks_per_col * bits / 8)
+    total_zp_bits = len(zero_points) * 8
+    mask = (1 << bits) - 1
+
+    for n in range(N):
+        unpacked_row_buf = []
+        current_bit_pos = 0
+        while current_bit_pos < total_zp_bits and len(unpacked_row_buf) < n_blocks_per_col:
+            byte_pos = (current_bit_pos // 8) + (n * zp_bits_per_n)
+            bit_offset = current_bit_pos % 8
+
+            bits_available = 8 - bit_offset
+            if bits_available >= bits:
+                value = (zero_points[byte_pos] >> bit_offset) & mask
+            else:
+                lower_bits = (zero_points[byte_pos] >> bit_offset) & mask
+                upper_bits = (zero_points[byte_pos + 1] << bits_available) & mask
+                value = upper_bits | lower_bits
+            unpacked_row_buf.append(value)
+            current_bit_pos += bits
+        unpacked_zp.extend(unpacked_row_buf)
+
+    return np.array(unpacked_zp, dtype=np.uint8)
+
+def matmulnbits_unpack_B(
+    B: np.ndarray,
+    N: int,
+    K: int,
+    n_blocks_per_col: int,
+    bits: int,
+    block_size: int
+) -> np.ndarray:
+    total_bits = n_blocks_per_col * math.ceil((block_size * bits) / 8) * 8
+    mask = (1 << bits) - 1
+    unpacked_B = np.empty((N, K), dtype=np.uint8)
+    for n in range(N):
+        unpacked_row_buf = []
+        current_bit_pos = 0
+        for n_bpc in range(n_blocks_per_col):
+            unpacked_col_buf = []
+            while len(unpacked_col_buf) < block_size and (len(unpacked_row_buf) + len(unpacked_col_buf)) < K and current_bit_pos < total_bits:
+                byte_pos = (current_bit_pos // 8)
+                bit_offset = current_bit_pos % 8
+
+                bits_available = 8 - bit_offset
+                if bits_available >= bits:
+                    value = (B[n][byte_pos] >> bit_offset) & mask
+                else:
+                    lower_bits = (B[n][byte_pos] >> bit_offset) & mask
+                    upper_bits = (B[n][byte_pos + 1] << bits_available) & mask
+                    value = upper_bits | lower_bits
+                unpacked_col_buf.append(value)
+                current_bit_pos += bits
+            unpacked_row_buf.extend(unpacked_col_buf)
+        unpacked_B[n] = unpacked_row_buf
```

Comment:
Suggestion: Deepcopy assignment of objects.

Python is finicky in maintaining scopes of objects, hence can change the object which are already assigned. Below example code will avoid such issues. This not true for this code, hence putting as a suggestion.

```python
from copy import deepcopy as cpy
unpacked_B[n] = cpy(unpacked_row_buf)
```

Reply:
Since this is not true for this code I think I will leave the code as is.